File: microgamer.ts
//% icon="\uf11b"
//% color="255" weight="100"
namespace MicroGamer {

    /**
     * CPP initialises the MicroGamer board
     */
    //% shim=MicroGamer::init_terminal
    export function init_terminal(): void {
        return;
    }

   /**
     * initialises the MicroGamer board
     */
    //% blockId=oled_init_terminal
    //% block="initialize MicroGamer board"
    //% icon="\uf1ec" 
    export function init(): void {
        init_terminal();
        pins.analogSetPitchPin(AnalogPin.P2)
        return;
    }

   /**
     * Update the display
     */
    //% subcategory=Display
    //% blockId=update_display
    //% block="Update display"
    //% icon="\uf1ec" 
    //% shim=MicroGamer::updateDisplay
    export function updateDisplay(): void {
        return;
    }

   /**
     * Disable display auto-update
     */
    //% subcategory=Display
    //% blockId=disable_display_autoupdate
    //% block="Disable display auto-update"
    //% icon="\uf1ec" 
    //% shim=MicroGamer::disableAutoUpdate
    export function disableAutoUpdate(): void {
        return;
    }

   /**
     * Enable display auto-update
     */
    //% subcategory=Display
    //% blockId=enable_display_autoupdate
    //% block="Enable display auto-update"
    //% icon="\uf1ec" 
    //% shim=MicroGamer::enableAutoUpdate
    export function enableAutoUpdate(): void {
        return;
    }

    /**
     * clears the display
     */
    //% subcategory=Display
    //% blockId=oled_clear_screen
    //% block="clear display"
    //% icon="\uf1ec" 
    //% shim=MicroGamer::clearDisplay
    export function clear(): void {
        return;
    }

    /**
     * prints a string on the MicroGamer display
     * @param text text to display
     */
    //% subcategory=Display
    //% weight=87 blockGap=8
    //% block="show|string %text" 
    //% blockId=oled_print_string
    //% icon="\uf1ec"
    //% shim=MicroGamer::showString
    export function showString(text: string): void {
        console.log("display: " + text);
        return;
    }

    /**
     * prints a number on the MicroGamer display
     * @param number number to display
     */
    //% subcategory=Display
    //% weight=96
    //% blockId=oled_print_number
    //% block="show|number %number" blockGap=8
    //% shim=MicroGamer::showNumber
    export function showNumber(number: number): void {
        console.log("display: " + number);
        return;
    }

    /**
     * Draw a pixel on the MicroGamer display
     * @param x X coordinate of the pixel
     * @param y Y coordinate of the pixel
     */
    //% subcategory=Display
    //% weight=96
    //% blockId=oled_draw_pixel
    //% block="draw pixel x %x| y %y|" blockGap=8
    //% shim=MicroGamer::drawPixel
    export function drawPixel(x: number, y: number): void {
        return;
    }

    /**
     * Draw a circle on the MicroGamer display
     * @param x X coordinate of the center
     * @param y Y coordinate of the center
     * @param r radius of the circle
     */
    //% subcategory=Display
    //% weight=96
    //% blockId=oled_draw_circle
    //% block="draw circle x %x| y %y| radius %radius|" blockGap=8
    //% shim=MicroGamer::drawCircle
    export function drawCircle(x: number, y: number, r:number): void {
        return;
    }

    /**
     * Fill a circle on the MicroGamer display
     * @param x X coordinate of the center
     * @param y Y coordinate of the center
     * @param r radius of the circle
     */
    //% subcategory=Display
    //% weight=96
    //% blockId=oled_fill_circle
    //% block="fill circle x %x| y %y| radius %radius|" blockGap=8
    //% shim=MicroGamer::fillCircle
    export function fillCircle(x: number, y: number, r:number): void {
        return;
    }

    /**
     * Draw a line on the MicroGamer display
     * @param x1 X coordinate of the first point
     * @param y1 Y coordinate of the first point
     * @param x2 X coordinate of the second point
     * @param y2 Y coordinate of the second point
     * @param r radius of the circle
     */
    //% subcategory=Display
    //% weight=96
    //% blockId=oled_draw_line
    //% block="draw line from x1 %x1| y1 %y1| to  x2 %x2| y2 %y2| " blockGap=8
    //% shim=MicroGamer::drawLine
    export function drawLine(x1: number, y1: number, x2:number, y2:number): void {
        return;
    }

    /**
     * Draw a rectangle on the MicroGamer display
     * @param x X coordinate of the top left corner
     * @param y Y coordinate of the top left corner
     * @param width width of the rectangle
     * @param height height of the rectangle
     */
    //% subcategory=Display
    //% weight=96
    //% blockId=oled_draw_rect
    //% block="draw rectangle x %x| y %y| width %width| height %height| " blockGap=8
    //% shim=MicroGamer::drawRect
    export function drawRect(x: number, y: number, width:number, height:number): void {
        return;
    }

    /**
     * Fill a rectangle on the MicroGamer display
     * @param x X coordinate of the top left corner
     * @param y Y coordinate of the top left corner
     * @param width width of the rectangle
     * @param height height of the rectangle
     */
    //% subcategory=Display
    //% weight=96
    //% blockId=oled_fill_rect
    //% block="fill rectangle x %x| y %y| width %width| height %height| " blockGap=8
    //% shim=MicroGamer::fillRect
    export function fillRect(x: number, y: number, width:number, height:number): void {
        return;
    }

    //  Button handling from:
    //   https://github.com/KitronikLtd/pxt-kitronik-zip-64/blob/master/zip64.ts
    // Copyright (c) 2018 Kitronik Ltd 

    /**
     * MicroGamer Button Pins
     */
    export enum MGButtonPins {
        //% block="Up"
        Up = <number>DAL.MICROBIT_ID_IO_P15,
        //% block="Down"
        Down = <number>DAL.MICROBIT_ID_IO_P14,
        //% block="Left"
        Left = <number>DAL.MICROBIT_ID_IO_P16,
        //% block="Right"
        Right = <number>DAL.MICROBIT_ID_IO_P13,
        //% block="X"
        X = <number>DAL.MICROBIT_ID_IO_P0,
        //% block="Y"
        Y = <number>DAL.MICROBIT_ID_IO_P1
    }


     /**
      * MicroGamer Button Events
     */
     export enum MGButtonEvents {
        //% block="down"
        Down = DAL.MICROBIT_BUTTON_EVT_DOWN,
        //% block="up"
        Up = DAL.MICROBIT_BUTTON_EVT_UP,
        //% block="click"
        Click = DAL.MICROBIT_BUTTON_EVT_CLICK
     }

    /**
     * Determines if a MicroGamer button is pressed
     * @param button press to be checked
     */
    //% subcategory=Inputs
    //% blockId="mg_ispressed" block="button %button|is pressed"
    //% button.fieldEditor="gridpicker" button.fieldOptions.columns=3
    //% weight=95 blockGap=8
    export function buttonIsPressed(button: MGButtonPins): boolean {
        const pin = <DigitalPin><number>button;
        pins.setPull(pin, PinPullMode.PullUp);
        return pins.digitalReadPin(pin) == 0;
    }

    /**
     * Do something when one of the MicroGamer Buttons is pressed
     * @param button press to be checked
     * @param event happening on the button, eg: click
     */
    //% subcategory=Inputs
    //% blockId="mg_button_press_on_event" block="on button %button|press %event"
    //% button.fieldEditor="gridpicker" button.fieldOptions.columns=3
    //% weight=93 blockGap=8
    export function onButtonPress(button: MGButtonPins, event: MGButtonEvents, handler: Action) {
        init();
        control.onEvent(<number>button, <number>event, handler);
    }

}

File: microgamer.cpp
#include "pxt.h"
#include "Adafruit_SSD1306.h"
using namespace pxt;

namespace MicroGamer {
    #define SSD1306_ADDRESS 0x78
    #undef printf
    #define WIDTH (128)
    #define HEIGHT (64)

    MicroBitI2C i2c(I2C_SDA0, I2C_SCL0);
    Adafruit_SSD1306_I2c *oled;
    bool auto_update = true;

    //%
    void init_terminal(){
        if (oled != NULL) {
            delete oled;
        }
        oled = new Adafruit_SSD1306_I2c(i2c, SSD1306_ADDRESS, HEIGHT, WIDTH);
        oled->clearDisplay();
        oled->display();
        oled->setTextCursor(0, 0);

        // From https://github.com/KitronikLtd/pxt-kitronik-zip-64/blob/master/zip64.cpp
        // Copyright (c) 2018 Kitronik Ltd 
        // This function sets all the buttons to actually appear as buttons on the micro:bit
#define ALLOC_PIN_BUTTON(id) new MicroBitButton(getPin(id)->name, id, MICROBIT_BUTTON_ALL_EVENTS, PullUp);
        ALLOC_PIN_BUTTON(MICROBIT_ID_IO_P0)
        ALLOC_PIN_BUTTON(MICROBIT_ID_IO_P1)
        ALLOC_PIN_BUTTON(MICROBIT_ID_IO_P13)
	ALLOC_PIN_BUTTON(MICROBIT_ID_IO_P14)
	ALLOC_PIN_BUTTON(MICROBIT_ID_IO_P15)
	ALLOC_PIN_BUTTON(MICROBIT_ID_IO_P16)
#undef ALLOC_PIN_BUTTON
    }

    //%
    void disableAutoUpdate(){
        auto_update = false;
    }

    //%
    void enableAutoUpdate(){
        auto_update = true;
    }

    //%
    void updateDisplay(){
        oled->display();
    }

    //%
    void showString(StringData *text) {
        oled->printf("%s\n", text->data);
        if (auto_update) {
            updateDisplay();
        }
    }

    //%
    void showNumber (int number) {
        oled->printf("%d\n", number);
        if (auto_update) {
            updateDisplay();
        }
    }

    //%
    void clearDisplay(){
        oled->clearDisplay();
        if (auto_update) {
            updateDisplay();
        }
    }

    //%
    void drawPixel(int x, int y){
        oled->drawPixel(x, y, WHITE);
        if (auto_update) {
            updateDisplay();
        }
    }

    //%
    void drawCircle(int x, int y, int r){
        oled->drawCircle(x, y, r, WHITE);
        if (auto_update) {
            updateDisplay();
        }
    }

    //%
    void fillCircle(int x, int y, int r){
        oled->fillCircle(x, y, r, WHITE);
        if (auto_update) {
            updateDisplay();
        }
    }

    //%
    void drawLine(int x0, int y0, int x1, int y1){
        oled->drawLine(x0, y0, x1, y1, WHITE);
        if (auto_update) {
            updateDisplay();
        }
    }

    //%
    void fillRect(int x, int y, int w, int h){
        oled->fillRect(x, y, w, h, WHITE);
         if (auto_update) {
            updateDisplay();
        }
   }

    //%
    void drawRect(int x, int y, int w, int h){
        oled->drawRect(x, y, w, h, WHITE);
        if (auto_update) {
            updateDisplay();
        }
    }

    //%
    void fillRoundRect(int x, int y, int w, int h, int r){
        oled->fillRoundRect(x, y, w, h, r, WHITE);
        if (auto_update) {
            updateDisplay();
        }
    }

    //%
    void drawRoundRect(int x, int y, int w, int h, int r){
    	oled->drawRoundRect(x, y, w, h, r, WHITE);
        if (auto_update) {
            updateDisplay();
        }
    }

    //%
    void drawTriangle(int x0, int y0, int x1, int y1, int x2, int y2){
    	oled->drawTriangle(x0, y0, x1, y1, x2, y2, WHITE);
        if (auto_update) {
            updateDisplay();
        }
    }

    //%
    void fillTriangle(int x0, int y0, int x1, int y1, int x2, int y2){
    	oled->fillTriangle(x0, y0, x1, y1, x2, y2, WHITE);
        if (auto_update) {
            updateDisplay();
        }
    }

    #define printf(...) uBit.serial.printf(__VA_ARGS__)

}
